CIP-????/README.md

```diff
@@ -0,0 +1,476 @@
+
+whereas we could do this:
+
+```
+(case (constr 1 x y) (alt x t1) (alt x y t2))
+```
+
+The effect of this is that we know statically that we have a function, and we can jump right into evaluating the _body_ of the case branch instead of first having to evaluate it to a function and apply the arguments.
+
+This saves us a small but meaningful amount of overhead at the cost of making the implementation significantly more complex.
+
+We prototyped this version and it was noticeably faster (~10%).
+However, performance investigations showed that we can realize a significant amount of the performance gain through other means, leaving only about 3% unclaimed.[^see-appendix-2]
+We think that this is an acceptable cost for a simpler implementation.
+
+[^see-appendix-2]: See Appendix 2 for more details.
+
+## Path to Active
+
+### Acceptance Criteria
+
+- [ ] Implementation in `plutus`, including specification and costing of the new operations
+- [ ] Further benchmarking with additional real-world examples
+- [ ] `plutus` changes integrated into `cardano-ledger`, including specification
+
+### Implementation Plan
+
+This plan will be implemented by Michael Peyton Jones and the Plutus Core team with assistance from the Ledger team.
+
+A prototype implementation already exists and is functional.
+
+## Appendices
+
+### Appendix 1: Changes to Typed Plutus Core
+
+While Typed Plutus Core is not part of the specification of Cardano, it is still interesting and informative to give the changes here.
+Plutus Core was originally conceived of as a typed language, and making sure that we can express the changes cleanly in a typed setting means that we ensure that the semantics make sense and that it will continue to be easy to compile to Plutus Core from typed languages.
+
+We add the following new type constructors to the type language of Typed Plutus Core:
+```
+(sum tys)
+(prod tys)
+```
+These correspond to sum and product types.
+While we are not using sums-and-products at the term level, we can keep it at the type level since that does not affect evaluation.
+
+We add the following new term constructors to the Typed Plutus Core language:
+```
+(constr ty i ts)
+(case ty t cs)
+```
+
+These are identical to their untyped cousins, except that they include a type annotation for the type of the whole term. 
+These make the typing rules much simpler, as otherwise we lack enough information to pin down the whole type.
+
+The typing rules for the new terms are:
+
+$$
+\frac{\Gamma \vdash t_i : ty_i,\ rty = (\mathrm{sum\ s_1 \ldots s_e \ldots s_m}), s_e = (\mathrm{prod}\ ty_1 \ldots ty_n)}{\Gamma \vdash (\mathrm{constr}\ rty\ e\ t_1 \ldots t_n) : rty}
+$$
+
+$$
+\frac{\Gamma \vdash t : tty,\ tty = (\mathrm{sum}\ s_1 \ldots s_n), s_i = (\mathrm{prod}\ p_{i_1} \ldots p_{i_m}),\ \Gamma \vdash c_i : p_{i_1} \rightarrow \ldots \rightarrow p_{i_m} \rightarrow rty}{\Gamma \vdash (\mathrm{case}\ rty\ t\ c_1 \ldots c_n) : rty}
+$$
```

Comment:
Do you need to put `rty` into the `case`? It's inferrable as long as there's a non-zero number of branches. Which begs another question: is zero number of branches allowed?

Reply:
I think it could be "allowed", but you just won't ever be able to evaluate it, because whatever `constr` value you look at, you won't have a branch for the right tag.

Reply:
Also I didn't answer the first question. Yes, you can infer it, but it's annoying: you have to go over each of the cases, collect all the result types, assert that they're the same, and then you have the type. Whereas it's much easier to check: you can directly construct the expected type of each case branch and then check that it has that type. And since this is TPLC it seemed fine to have slightly more type annotations to make it simpler to check things.

(Interestingly GHC Core also annotates case expressions with the result type!)

Reply:
> Yes, you can infer it, but it's annoying: you have to go over each of the cases, collect all the result types, assert that they're the same, and then you have the type. Whereas it's much easier to check: you can directly construct the expected type of each case branch and then check that it has that type. And since this is TPLC it seemed fine to have slightly more type annotations to make it simpler to check things.

Right, great points. Would be worth adding them to a Note when `Case` gets into the AST.

Reply:
Yes, but I think not quite interesting enough to go in the CIP.